namesys/routing.go

```diff
@@ -98,10 +130,35 @@ func (r *routingResolver) resolveOnce(ctx context.Context, name string) (path.Pa
 	valh, err := mh.Cast(entry.GetValue())
 	if err != nil {
 		// Not a multihash, probably a new record
-		return path.ParsePath(string(entry.GetValue()))
+		p, err := path.ParsePath(string(entry.GetValue()))
+		if err != nil {
+			return "", err
+		}
+
+		cacheTil := time.Now().Add(r.cachelife)
+		eol, ok := checkEOL(entry)
+		if ok && eol.Before(cacheTil) {
+			cacheTil = eol
+		}
```

Comment:
this stuff could get easily wrong, and cache time errors could behavior drastically. i'd propose feeding `entry` into `r.cacheSet` and put the computation in one place, calculated always from the original data, instead of intermediate values.